app/controllers/owned_tag_sets_controller.rb

```diff
@@ -74,31 +74,31 @@ def show
       # we use this to store the tag name results
       @tag_hash = HashWithIndifferentAccess.new
 
-      %w(character relationship).each do |tag_type|
-        if @tag_set.has_type?(tag_type)
-          ## names_by_parent returns a hash of arrays like so:
-          ## hash[parent_name] => [child name, child name, child name]
-
-          # get the manually associated fandoms
-          assoc_hash = TagSetAssociation.names_by_parent(TagSetAssociation.for_tag_set(@tag_set), tag_type)
-
-          # get canonically associated fandoms
-          # Safe for constantize as tag_type restricted to character relationship
-          canonical_hash = Tag.names_by_parent(tag_type.classify.constantize.in_tag_set(@tag_set), "fandom")
-
-          # merge the values of the two hashes (each value is an array) as a set (ie remove duplicates)
-          @tag_hash[tag_type] = assoc_hash.merge(canonical_hash) {|key, oldval, newval| (oldval | newval) }
-
-          # get any tags without a fandom
-          remaining = @tag_set.with_type(tag_type).with_no_parents
-          if remaining.count > 0
-            @tag_hash[tag_type]["(No linked fandom - might need association)"] ||= []
-            @tag_hash[tag_type]["(No linked fandom - might need association)"] += remaining.pluck(:name)
-          end
+      %w[character relationship].each do |tag_type|
+        next unless @tag_set.has_type?(tag_type)
+
+        ## names_by_parent returns a hash of arrays like so:
+        ## hash[parent_name] => [child name, child name, child name]
+
+        # get the manually associated fandoms
+        assoc_hash = TagSetAssociation.names_by_parent(TagSetAssociation.for_tag_set(@tag_set), tag_type)
 
-          # store the parents
-          @fandom_keys_from_other_tags += @tag_hash[tag_type].keys
+        # get canonically associated fandoms
+        # Safe for constantize as tag_type restricted to character relationship
+        canonical_hash = Tag.names_by_parent(tag_type.classify.constantize.in_tag_set(@tag_set).canonical, "fandom")
+
+        # merge the values of the two hashes (each value is an array) as a set (ie remove duplicates)
+        @tag_hash[tag_type] = assoc_hash.merge(canonical_hash) { |_key, oldval, newval| (oldval | newval) }
+
+        # get any tags without a fandom
+        remaining = @tag_set.with_type(tag_type).where.not(name: @tag_hash[tag_type].values.flatten)
```

Comment:
This new code make sense to me but help me a bit - how did the old code avoid grabbing tags that have manual associations? In the new one it's clear, they are included in the merged hash. But in the old one..?

Reply:
I don't think the old code _would_ have done that properly. The only reason it _seemed_ to work (in this case) was that we're dealing with non-canonical tags that did have a parent, so regardless of whether there was a manual association, they wouldn't show up in `with_no_parents`. If someone had an association for a fandom that didn't have a `CommonTagging` yet to that relationship/character, I think the old code would not have done the right thing. I don't know how common that is, though